lib/Doctrine/DBAL/Types/ArrayOf.php

```diff
@@ -0,0 +1,31 @@
+<?php
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Types;
+
+use Doctrine\DBAL\Connection;
+
+/**
+ * Type representing an array of an other type, not an actual type even though it should be but that would require to
+ * rewrite too much code because types are not allowed to rewrite the query and replace its parameters.
+ */
+class ArrayOf
```

Comment:
I don't think it's a good idea. This way, the DBAL API still won't enforce the argument types. A better approach might be to make the `bind*` methods typed. See https://github.com/doctrine/dbal/issues/4602#issuecomment-826979381.

For this specific issue, we can just introduce `bindArray()` with the same signature as `bindValue()`:
https://github.com/doctrine/dbal/blob/de7f86d1ce98d32cb1c3cc36703b3182e44f74d7/lib/Doctrine/DBAL/Statement.php#L101

The only difference will be that `$value` will always have to be an array.

Reply:
I agree with you, this should be done this way in a future release, typically the 3.2.x.
But I think my approach solves the problem by introducing absolutely no BC-break because it keeps the existing parameter expansion logic.

It could also be done without a new object, typically by just using a naming convention like `datetime[]` and testing the existence of `[]` at the end of the type but I think a new wrapper class is simple and better.

Reply:
> Since this is a new feature, it will have to target `3.2.x`.

I'm sorry but I have to disagree with you, this is clearly a bugfix although it introduces a new "type" it allows to fix an issue that was here since the beginning of this project and introduces no BC-break.

That being said, my initial goal was to fix this issue in doctrine/orm but now that I saw the way the UnitOfWork deals with identifiers (imploding them like they're all strings) and more generally with types in queries, I must say I'm not eager to touch anything in this mess. So if you still want to rebase this in 3.2.x I'm fine with that.

I've already rolled back to not using custom object types in my projects anyway.